controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,448 @@
+	}
+
+	// marked for deletion
+	if !r.clientMappingConfig.GetDeletionTimestamp().IsZero() {
+		if res, err := r.disableBlockPoolMirroring(); err != nil || !res.IsZero() {
+			return res, err
+		}
+
+		if controllerutil.RemoveFinalizer(r.clientMappingConfig, mirroringFinalizer) {
+			r.log.Info(
+				"removing finalizer from ClientMappingConfig.",
+				"ClientMappingConfig",
+				r.clientMappingConfig.Name,
+			)
+			if err := r.update(r.clientMappingConfig); err != nil {
+				r.log.Info(
+					"Failed to remove finalizer from ClientMappingConfig",
+					"ClientMappingConfig",
+					r.clientMappingConfig.Name,
+				)
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from ClientMappingConfig: %v", err)
+			}
+		}
+	}
+
+	if len(r.clientMappingConfig.Data) < 1 {
+		return ctrl.Result{}, nil
+	}
+
+	if controllerutil.AddFinalizer(r.clientMappingConfig, mirroringFinalizer) {
+		r.log.Info(
+			"Finalizer not found for ClientMappingConfig. Adding finalizer.",
+			"ClientMappingConfig",
+			r.clientMappingConfig.Name,
+		)
+		if err := r.update(r.clientMappingConfig); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update ClientMappingConfig: %v", err)
+		}
+	}
+
+	// Find the StorageClusterPeer from OwnerRef
+	owner := util.FindOwnerRefByKind(r.clientMappingConfig, "StorageClusterPeer")
+	if owner == nil {
+		return ctrl.Result{}, fmt.Errorf("failed to find StorgeClusterPeer owning the ClientMappingConfig")
+	}
+
+	// Fetch the StorageClusterPeer instance
+	r.storageClusterPeer = &ocsv1.StorageClusterPeer{}
+	r.storageClusterPeer.Name = owner.Name
+	r.storageClusterPeer.Namespace = request.Namespace
+
+	if err = r.get(r.storageClusterPeer); err != nil {
+		if k8serrors.IsNotFound(err) {
+			r.log.Info("StorageClusterPeer resource not found. Ignoring since object must be deleted.")
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClusterPeer.")
+		return ctrl.Result{}, err
+	}
+
+	if r.storageClusterPeer.Status.State != ocsv1.StorageClusterPeerStatePeered {
+		return ctrl.Result{}, fmt.Errorf(
+			"waiting for StorageClusterPeer %s to be in Peered state",
+			r.storageClusterPeer.Name,
+		)
+	}
+
+	return r.reconcilePhases()
+}
+
+func (r *MirroringReconciler) reconcilePhases() (ctrl.Result, error) {
+	ocsClient, err := providerClient.NewProviderClient(r.ctx, r.storageClusterPeer.Spec.ApiEndpoint, time.Second*10)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create a new provider client: %v", err)
+	}
+	defer ocsClient.Close()
+
+	if err = r.reconcileRBDMirrorDaemon(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if res, err := r.reconcileBlockPoolMirroring(ocsClient); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if res, err := r.reconcileRadosNamespaceMirroring(ocsClient); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) disableBlockPoolMirroring() (ctrl.Result, error) {
```

Comment:
Can't this be just another value of the normal reconciliation of the block pool?